# Question 1
On Shopify, we have exactly 100 sneaker shops, and each of these shops sells only one model of shoe. We want to do some analysis of the average order value (AOV). When we look at orders data over a 30 day window, we naively calculate an AOV of $3145.13. Given that we know these shops are selling sneakers, a relatively affordable item, something seems wrong with our analysis. 

In [1]:
#Import Packages
import pandas as pd
import numpy as np

In [2]:
#Import Data
data = pd.read_csv("./2019 Winter Data Science Intern Challenge Data Set - Sheet1.csv")
print("data size: ", len(data))
print("sample data:")
data.head()

data size:  5000
sample data:


order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11

In [3]:
print("This is the AOV in the question:", data['order_amount'].mean())
print("This is the max order_amount in this dataset", data['order_amount'].max())
print("This is the lines that are much more larger than the original average:")
outliner = data[data["order_amount"]>data['order_amount'].mean()]
print(len(outliner))
outliner

This is the AOV in the question: 3145.128
This is the max order_amount in this dataset 704000
This is the lines that are much more larger than the original average:
63


order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
160        161       78      990         25725            1    credit_card   
490        491       78      936         51450            2          debit   
493        494       78      983         51450            2           cash   
...        ...      ...      ...           ...          ...            ...   
4646      4647       42      607        704000         2000    credit_card   
4715      4716       78      818         77175            3          debit   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   
4918      4919       78      823         25725            1           cash   

               created_at  
15     2017-03-07 4:00:00  
60     2017-03-04 4:00:00  
160    2017-03-12 5:56:57  
490   2017-03-26 17:08:19  
493   2017-03-16 21:39:35  
...                   ...  
4646   2017-03-02 4:00:00  
4715   2017-03-05 5:10:44  
4868   2017-03-22 4:00:00  
4882   2017-03-25 4:00:00  
4918  2017-03-15 13:26:46  

[63 rows x 7 columns]

In [4]:
#To show where the outliners come from
print(outliner['order_amount'].unique())
print(outliner['shop_id'].unique())

[704000  25725  51450 154350  77175 102900]
[42 78]


In [5]:
print(outliner['order_amount']/outliner['total_items'])

15        352.0
60        352.0
160     25725.0
490     25725.0
493     25725.0
         ...   
4646      352.0
4715    25725.0
4868      352.0
4882      352.0
4918    25725.0
Length: 63, dtype: float64


In [6]:
print(data[data["order_amount"]==704000])
# We found the orders about outliner 704000 all come from the same shop and same user, 
# so we can considered as tests or frauds

      order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
520        521       42      607        704000         2000    credit_card   
1104      1105       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1436      1437       42      607        704000         2000    credit_card   
1562      1563       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
2153      2154       42      607        704000         2000    credit_card   
2297      2298       42      607        704000         2000    credit_card   
2835      2836       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    c

In [7]:
outliner[outliner["order_amount"]!=704000]

order_id  shop_id  user_id  order_amount  total_items payment_method  \
160        161       78      990         25725            1    credit_card   
490        491       78      936         51450            2          debit   
493        494       78      983         51450            2           cash   
511        512       78      967         51450            2           cash   
617        618       78      760         51450            2           cash   
691        692       78      878        154350            6          debit   
1056      1057       78      800         25725            1          debit   
1193      1194       78      944         25725            1          debit   
1204      1205       78      970         25725            1    credit_card   
1259      1260       78      775         77175            3    credit_card   
1384      1385       78      867         25725            1           cash   
1419      1420       78      912         25725            1           cash   
1452      1453       78      812         25725            1    credit_card   
1529      1530       78      810         51450            2           cash   
2270      2271       78      855         25725            1    credit_card   
2452      2453       78      709         51450            2           cash   
2492      2493       78      834        102900            4          debit   
2495      2496       78      707         51450            2           cash   
2512      2513       78      935         51450            2          debit   
2548      2549       78      861         25725            1           cash   
2564      2565       78      915         77175            3          debit   
2690      2691       78      962         77175            3          debit   
2773      2774       78      890         25725            1           cash   
2818      2819       78      869         51450            2          debit   
2821      2822       78      814         51450            2           cash   
2906      2907       78      817         77175            3          debit   
2922      2923       78      740         25725            1          debit   
3085      3086       78      910         25725            1           cash   
3101      3102       78      855         51450            2    credit_card   
3151      3152       78      745         25725            1    credit_card   
3167      3168       78      927         51450            2           cash   
3403      3404       78      928         77175            3          debit   
3440      3441       78      982         25725            1          debit   
3705      3706       78      828         51450            2    credit_card   
3724      3725       78      766         77175            3    credit_card   
3780      3781       78      889         25725            1           cash   
4040      4041       78      852         25725            1           cash   
4079      4080       78      946         51450            2           cash   
4192      4193       78      787         77175            3    credit_card   
4311      4312       78      960         51450            2          debit   
4412      4413       78      756         51450            2          debit   
4420      4421       78      969         77175            3          debit   
4505      4506       78      866         25725            1          debit   
4584      4585       78      997         25725            1           cash   
4715      4716       78      818         77175            3          debit   
4918      4919       78      823         25725            1           cash   

               created_at  
160    2017-03-12 5:56:57  
490   2017-03-26 17:08:19  
493   2017-03-16 21:39:35  
511    2017-03-09 7:23:14  
617   2017-03-18 11:18:42  
691   2017-03-27 22:51:43  
1056  2017-03-15 10:16:45  
1193  2017-03-16 16:38:26  
1204  2017-03-17 22:32:21  
1259   2017-03-27 9:27:20  
1384  2017-03-17 16:38:06  
141

In [8]:
# Remove outliners
data_2 = data[data["order_amount"]<=data['order_amount'].mean()]
print("size of dataset after removing the outliners:", len(data_2))
print("sample in data_2:")
data_2.head()

size of dataset after removing the outliners: 4937
sample in data_2:


order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11

In [9]:
#recalculate the AOV
print("new AOV is:", data_2['order_amount'].mean())

new AOV is: 302.58051448247926


In [11]:
#for the orders which shop_id = 78
shop_78 = outliner[outliner['shop_id'] == 78]
print("AOV for shop_id = 78:", shop_78['order_amount'].mean())
shop_78.head()

AOV for shop_id = 78: 49213.04347826087


order_id  shop_id  user_id  order_amount  total_items payment_method  \
160       161       78      990         25725            1    credit_card   
490       491       78      936         51450            2          debit   
493       494       78      983         51450            2           cash   
511       512       78      967         51450            2           cash   
617       618       78      760         51450            2           cash   

              created_at  
160   2017-03-12 5:56:57  
490  2017-03-26 17:08:19  
493  2017-03-16 21:39:35  
511   2017-03-09 7:23:14  
617  2017-03-18 11:18:42

## a. Think about what could be going wrong with our calculation. Think about a better way to evaluate this data.  
Answer: The problem happened because: 1.The shop with shop_id = 78 contains the item which value is 25725 which is much higher that items in other shops; 2. The shop with shop_id = 42 has so many similar orders from the same customer whose user_id = 607, we can consider the actions as tests or click farming.  
So what we should do is to remove the outliners, for the shop_id = 42, we just simply remove the orders from our dataset, but for the shop_id = 78, we should use another single table to show its AOV in order to have a comprehensive analysis of AOV in 30-days window. And then calculate the AOV after removing outliners or we can calculate Median Order Value instead. In this case, I will choose the first solution.

## b. What metric would you report for this dataset?
Answer: AOV after removing the outliners.  

## c. What is its value?
Answer: 302.58


# Question 2

## a. How many orders were shipped by Speedy Express in total?
54

In [ ]:
SELECT COUNT(OrderID) 
FROM [Orders] 
LEFT JOIN [Shippers] on [Orders].ShipperID == [Shippers].ShipperID
WHERE ShipperName == "Speedy Express"

## b.What is the last name of the employee with the most orders?
Peacock

In [ ]:
SELECT LastName FROM
(SELECT EmployeeID, COUNT(*) as count 
FROM [Orders]
GROUP BY EmployeeID
ORDER BY count DESC
LIMIT 1) AS Temp
LEFT JOIN [Employees] on [Temp].EmployeeID == [Employees].EmployeeID

## c.What product was ordered the most by customers in Germany?
Boston Crab Meat

In [ ]:
SELECT ProductName
FROM
(SELECT ProductID, sum(Quantity) as Sum FROM [Orders]
LEFT JOIN [OrderDetails] on [Orders].OrderID == [OrderDetails].OrderID
LEFT JOIN [Customers] on [Orders].CustomerID == [Customers].CustomerID
WHERE Country == 'Germany'
GROUP BY ProductID
Order By Sum desc
LIMIT 1) AS TEMP2
LEFT JOIN [Products] on [TEMP2].ProductID == [Products].ProductID